m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -0,0 +1,831 @@
+        }
+
+        int numOverheadBytes = EMIT_METRIC_BATCH_OVERHEAD + size;
+
+        freeBytes = maxPacketSizeBytes - numOverheadBytes;
+
+        if (freeBytes <= 0) {
+            throw new IllegalArgumentException("Common tags serialized size exceeds packet size");
+        }
+
+        return freeBytes;
+    }
+
+    private static String getHostName() {
+        try {
+            return InetAddress.getLocalHost().getHostName();
+        } catch (UnknownHostException e) {
+            LOG.warn("Unable to determine hostname. Defaulting to: %s", DEFAULT_TAG_VALUE);
+            return DEFAULT_TAG_VALUE;
+        }
+    }
+
+    private void addAndRunProcessor() {
+        phaser.register();
+
+        executor.execute(new Processor());
+    }
+
+    @Override
+    public Capabilities capabilities() {
+        return CapableOf.REPORTING_TAGGING;
+    }
+
+    @Override
+    public void flush() {
+        try {
+            metricQueue.put(SizedMetric.FLUSH);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+
+            throw new RuntimeException(e);
+        }
+    }
+
+    private void flush(List<Metric> metrics) {
+        flush(metrics, DEFAULT_FLUSH_TRY_COUNT);
+    }
+
+    private void flush(List<Metric> metrics, int maxTries) {
+        if (metrics.isEmpty()) {
+            return;
+        }
+
+        TException exception = null;
+
+        int tries = 0;
+        boolean tryFlush = true;
+
+        synchronized (metricBatch) {
+            metricBatch.setMetrics(metrics);
+
+            while (tryFlush) {
+                try {
+                    client.emitMetricBatch(metricBatch);
+
+                    tryFlush = false;
+                } catch (TException tException) {
+                    tries++;
+
+                    if (tries < maxTries) {
+                        try {
+                            Thread.sleep(FLUSH_RETRY_DELAY_MILLIS);
+                        } catch (InterruptedException interruptedException) {
+                            Thread.currentThread().interrupt();
+
+                            // If someone interrupts this thread, we should probably
+                            // not still try to flush
+                            tryFlush = false;
+
+                            exception = new TException("Interrupted while about to retry emitting metrics", tException);
+                        }
+                    } else {
+                        tryFlush = false;
+
+                        // Store exception to throw later after clean up.
+                        // Don't want to put everything in a finally block here,
+                        // which will hold the lock unnecessarily long
+                        exception = tException;
+                    }
+                }
+            }
+
+            metricBatch.setMetrics(null);
+        }
+
+        metrics.clear();
+
+        if (exception != null) {
+            throw new RuntimeException(
```

Comment:
per above, not sure we should be throwing exceptions

Reply:
Valid point. Will warn in log instead of throwing if it's possible to continue, but will stick to throwing exceptions when there is something fundamentally wrong, like a missing required tag or failing to open a socket.

Reply:
Sounds good. Ideally, we'd only throw an exception during scope construction.